src/inet/InetLayer.h

```diff
@@ -128,19 +122,92 @@ class DLL_EXPORT InetLayer
     void * GetPlatformData();
     void SetPlatformData(void * aPlatformData);
 
-#if INET_CONFIG_ENABLE_TCP_ENDPOINT && INET_TCP_IDLE_CHECK_INTERVAL > 0
-    static void HandleTCPInactivityTimer(chip::System::Layer * aSystemLayer, void * aAppState);
-#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT && INET_TCP_IDLE_CHECK_INTERVAL > 0
+protected:
+    InetLayer();
+    virtual ~InetLayer() {}
+
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    friend class TCPEndPointDeletor;
+
+    virtual TCPEndPoint * CreateTCPEndPoint(InetLayer & layer) = 0;
+    virtual void DeleteTCPEndPoint(TCPEndPoint * endPoint)     = 0;
+
+    using TCPEndPointVisitor                                          = bool (*)(TCPEndPoint *);
+    virtual bool ForEachTCPEndPoint(const TCPEndPointVisitor visitor) = 0;
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+
+#if INET_CONFIG_ENABLE_UDP_ENDPOINT
+    friend class UDPEndPointDeletor;
+    virtual UDPEndPoint * CreateUDPEndPoint(InetLayer & layer) = 0;
+    virtual void DeleteUDPEndPoint(UDPEndPoint * endPoint)     = 0;
+
+    using UDPEndPointVisitor                                         = bool (*)(UDPEndPoint *);
+    virtual bool ForEachUDPEndPoint(const UDPEndPointVisitor vistor) = 0;
+#endif // INET_CONFIG_ENABLE_UDP_ENDPOINT
 
 private:
     ObjectLifeCycle mLayerState;
     void * mContext;
     void * mPlatformData;
     chip::System::Layer * mSystemLayer;
+};
 
-    bool IsIdleTimerRunning();
+template <typename T, typename U>
+class InetLayerPoolImpl : public InetLayer
+{
+    using TCPEndPointType = T;
+    using UDPEndPointType = U;
+
+public:
+protected:
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    TCPEndPoint * CreateTCPEndPoint(InetLayer & layer) override { return sTCPEndPointPool.CreateObject(layer); }
+    void DeleteTCPEndPoint(TCPEndPoint * endPoint) override
+    {
+        sTCPEndPointPool.ReleaseObject(static_cast<TCPEndPointImpl *>(endPoint));
+    }
+    bool ForEachTCPEndPoint(const TCPEndPointVisitor visitor) override
+    {
+        return sTCPEndPointPool.ForEachActiveObject(
+            [&](TCPEndPoint * endPoint) -> bool { return (&endPoint->Layer() == this) ? visitor(endPoint) : true; });
+    }
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+
+#if INET_CONFIG_ENABLE_UDP_ENDPOINT
+    UDPEndPoint * CreateUDPEndPoint(InetLayer & layer) override { return sUDPEndPointPool.CreateObject(layer); }
+    void DeleteUDPEndPoint(UDPEndPoint * endPoint) override
+    {
+        sUDPEndPointPool.ReleaseObject(static_cast<UDPEndPointImpl *>(endPoint));
+    }
+    bool ForEachUDPEndPoint(const UDPEndPointVisitor visitor) override
+    {
+        return sUDPEndPointPool.ForEachActiveObject(
+            [&](UDPEndPoint * endPoint) -> bool { return (&endPoint->Layer() == this) ? visitor(endPoint) : true; });
+    }
+#endif // INET_CONFIG_ENABLE_UDP_ENDPOINT
+
+private:
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    static ObjectPool<TCPEndPointType, INET_CONFIG_NUM_TCP_ENDPOINTS> sTCPEndPointPool;
```

Comment:
Does it need to be static if it's anyway part of a global object?

Reply:
I haven't settled on this yet. One use of virtualized InetLayer is testing with multiple instances, not necessarily global, in a single binary, and I'm not sure whether or not we'd want a shared pool.

It is also the case the only virtual `InetLayer` methods are those involving endpoint allocation and iteration, so I'm also looking at whether it would be cleaner to split them off into endpoint factories.